File: frontend/src/api/scan.ts
/**
 * 전수 점검 API 클라이언트
 */

import axios from 'axios';

const API_URL = `http://${window.location.hostname}:8000`;

export type ScanType = 'scan-all' | 'scan' | 'scan-db';

export interface FullScanRequest {
  server_ids: string[];
  scan_type?: ScanType;
}

export interface FullScanResponse {
  job_id: string;
  message: string;
  total_servers: number;
  status: string;
}

export interface ScanProgress {
  job_id: string;
  status: 'queued' | 'running' | 'completed' | 'failed';
  progress: number;
  current_step: number;
  current_server?: string;
  completed_servers: number;
  total_servers: number;
  message: string;
}

export interface TopVulnerableServer {
  server_id: string;
  hostname: string;
  count: number;
}

export interface RiskDistribution {
  low: number;
  medium: number;
  high: number;
}

export interface ScanResult {
  job_id: string;
  company: string;
  total_servers: number;
  scan_duration: string;
  vulnerable_count: number;
  secure_count: number;
  risk_percentage: number;
  top_vulnerable_server?: TopVulnerableServer;
  risk_distribution: RiskDistribution;
  scan_completed_at: string;
}

/**
 * 전수 점검 시작
 */
export async function startFullScan(serverIds: string[], scanType: ScanType = 'scan-all'): Promise<FullScanResponse> {
  const token = localStorage.getItem('access_token');

  const response = await axios.post<FullScanResponse>(
    `${API_URL}/api/scan/full`,
    { server_ids: serverIds, scan_type: scanType },
    {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    }
  );

  return response.data;
}

/**
 * 점검 진행률 조회
 */
export async function getScanProgress(jobId: string): Promise<ScanProgress> {
  const token = localStorage.getItem('access_token');

  const response = await axios.get<ScanProgress>(
    `${API_URL}/api/scan/progress/${jobId}`,
    {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    }
  );

  return response.data;
}

/**
 * 점검 결과 조회
 */
export async function getScanResult(jobId: string): Promise<ScanResult> {
  const token = localStorage.getItem('access_token');

  const response = await axios.get<ScanResult>(
    `${API_URL}/api/scan/result/${jobId}`,
    {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    }
  );

  return response.data;
}

/**
 * 회사의 모든 서버 ID 조회
 */
export async function getAllServerIds(): Promise<string[]> {
  const token = localStorage.getItem('access_token');

  const response = await axios.get<any[]>(
    `${API_URL}/api/assets`,
    {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    }
  );

  return response.data.map((server: any) => server.server_id);
}

File: backend/schemas/scan.py
"""
전수 점검 Pydantic 스키마
"""

from pydantic import BaseModel, Field
from typing import List, Optional, Literal


class FullScanRequest(BaseModel):
    """전수 점검 요청"""
    server_ids: List[str] = Field(..., min_length=1, description="점검할 서버 ID 목록")
    scan_type: Literal["scan-all", "scan", "scan-db"] = Field(default="scan-all", description="점검 유형")


class ScanProgressResponse(BaseModel):
    """점검 진행 상황 응답"""
    job_id: str
    status: Literal["queued", "running", "completed", "failed"]
    progress: int = Field(ge=0, le=100, description="진행률 (0-100)")
    current_step: int = Field(ge=1, le=4, description="현재 단계 (1-4)")
    current_server: Optional[str] = None
    completed_servers: int
    total_servers: int
    message: str


class TopVulnerableServer(BaseModel):
    """최다 취약점 서버"""
    server_id: str
    hostname: str
    count: int


class RiskDistribution(BaseModel):
    """위험도 분포"""
    low: int = Field(description="저위험 비율 (%)")
    medium: int = Field(description="중위험 비율 (%)")
    high: int = Field(description="고위험 비율 (%)")


class ScanResultResponse(BaseModel):
    """점검 결과 요약"""
    job_id: str
    company: str
    total_servers: int
    scan_duration: str
    vulnerable_count: int
    secure_count: int
    risk_percentage: int
    top_vulnerable_server: Optional[TopVulnerableServer]
    risk_distribution: RiskDistribution
    scan_completed_at: str
